// @ensdomains/dnssecoraclejs-017/dist/typechain/factories/SHA1__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SHA1__factory = void 0;
const contracts_1 = require("@ethersproject/contracts");
class SHA1__factory extends contracts_1.ContractFactory {
    constructor(signer) {
        super(_abi, _bytecode, signer);
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static connect(address, signerOrProvider) {
        return new contracts_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.SHA1__factory = SHA1__factory;
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "bytes32",
                name: "x",
                type: "bytes32",
            },
        ],
        name: "Debug",
        type: "event",
    },
];
const _bytecode = "0x60566023600b82828239805160001a607314601657fe5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea264697066735822122099f1d2a1435aede1991b436fdf4f8015f0227f6ec3689db1c94d1d6917207e8964736f6c63430007040033";
