// @ensdomains/mock/dist/testing-utils/deployENS.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _deployDNSSEC = _interopRequireDefault(require("./deployDNSSEC"));

var _utils = require("./utils");

var _table = require("table");

var _namelogger = require("./namelogger");

var _interfaces = require("../constants/interfaces");

var ROOT_NODE = '0x00000000000000000000000000000000'; // ipfs://QmTeW79w7QQ6Npa3b1d5tANreCDxF2iDaAPsDvW6KtLmfB

var contenthash = '0xe301017012204edd2984eeaf3ddf50bac238ec95c5713fb40b5e428b508fdbe55d3b9f155ffe';
var content = '0x736f6d65436f6e74656e74000000000000000000000000000000000000000000';
var ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'; // dnslink based ipns'app.uniswap.org'

var deprecated_contenthash = '0xe5010170000f6170702e756e69737761702e6f7267';

var toBN = require('web3-utils').toBN;

var legacyRegistrarInterfaceId = _interfaces.interfaces.legacyRegistrar,
    permanentRegistrarInterfaceId = _interfaces.interfaces.permanentRegistrar,
    permanentRegistrarWithConfigInterfaceId = _interfaces.interfaces.permanentRegistrarWithConfig,
    bulkRenewalInterfaceId = _interfaces.interfaces.bulkRenewal,
    linearPriceOracleInterfaceId = _interfaces.interfaces.linearPriceOracle;

function deployENS(_x) {
  return _deployENS.apply(this, arguments);
}

function _deployENS() {
  _deployENS = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee5(_ref) {
    var _response;

    var web3, accounts, _ref$dnssec, dnssec, sha3, namehash, nameLogger, registryJSON, resolverJSON, oldResolverJSON, reverseRegistrarJSON, priceOracleJSON, linearPremiumPriceOracleJSON, dummyOracleJSON, controllerJSON, bulkRenewalJSON, testRegistrarJSON, legacyAuctionRegistrarSimplifiedJSON, ENSWithFallbackJSON, oldBaseRegistrarJSON, newBaseRegistrarJSON, registrarMigrationJSON, EthRegistrarSubdomainRegistrarJSON, ENSMigrationSubdomainRegistrarJSON, ens, resolver, oldResolver, oldReverseRegistrar, testRegistrar, eightweeks, startTime, legacyAuctionRegistrar, ensContract, resolverContract, oldResolverContract, oldReverseRegistrarContract, testRegistrarContract, legacyAuctionRegistrarContract, tld, tldHash, legacynames, i, lockoutlength, now, priceOracle, oldBaseRegistrar, controller, oldBaseRegistrarContract, controllerContract, newnames, aBitTooAwesome, aBitTooAwesome2, aBitTooAwesome3, otherOwner, addResolverAndRecords, _addResolverAndRecords, contractdomain, current, oldEns, label, oldSubdomainRegistrar, oldSubdomainRegistrarContract, subdomainRegistrar, newEns, newEnsContract, newBaseRegistrar, newBaseRegistrarContract, dummyOracleRate, dummyOracle, dummyOracleContract, latestAnswer, premium, decreaseDuration, decreaseRate, linearPriceOracle, linearPriceOracleContract, newController, newControllerContract, newResolver, newResolverContract, addNewResolverAndRecords, _addNewResolverAndRecords, bulkRenewal, newTestRegistrar, newReverseRegistrar, registrarMigration, registrarMigrationContract, newTestRegistrarContract, name, domain, labelhash, owner, tx, migrate, setNewResolver, _setNewResolver, baseDays, beforeTime, afterTime, sixcharprice, fourcharprice, threecharprice, response, config, contractNames, contractAddressesTable, output, labels;

    return _regenerator["default"].wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _setNewResolver = function _ref9() {
              _setNewResolver = (0, _asyncToGenerator2["default"])(
              /*#__PURE__*/
              _regenerator["default"].mark(function _callee4(name) {
                return _regenerator["default"].wrap(function _callee4$(_context4) {
                  while (1) {
                    switch (_context4.prev = _context4.next) {
                      case 0:
                        _context4.next = 2;
                        return newEnsContract.setResolver(namehash(name), newResolver._address).send({
                          from: accounts[0]
                        });

                      case 2:
                      case "end":
                        return _context4.stop();
                    }
                  }
                }, _callee4);
              }));
              return _setNewResolver.apply(this, arguments);
            };

            setNewResolver = function _ref8(_x6) {
              return _setNewResolver.apply(this, arguments);
            };

            _addNewResolverAndRecords = function _ref7() {
              _addNewResolverAndRecords = (0, _asyncToGenerator2["default"])(
              /*#__PURE__*/
              _regenerator["default"].mark(function _callee3(name) {
                var hash;
                return _regenerator["default"].wrap(function _callee3$(_context3) {
                  while (1) {
                    switch (_context3.prev = _context3.next) {
                      case 0:
                        console.log('setting up ', name);
                        hash = namehash(name);
                        console.log('resolver');
                        _context3.next = 5;
                        return newEnsContract.setResolver(hash, newResolver._address).send({
                          from: accounts[0]
                        });

                      case 5:
                        console.log('addr');
                        _context3.next = 8;
                        return newResolverContract.setAddr(hash, newResolver._address).send({
                          from: accounts[0]
                        });

                      case 8:
                        // ipfs://QmTeW79w7QQ6Npa3b1d5tANreCDxF2iDaAPsDvW6KtLmfB
                        console.log('contenthash');
                        _context3.next = 11;
                        return newResolverContract.setContenthash(hash, contenthash).send({
                          gas: 5000000,
                          from: accounts[0]
                        });

                      case 11:
                        console.log('finished setting up', name);

                      case 12:
                      case "end":
                        return _context3.stop();
                    }
                  }
                }, _callee3);
              }));
              return _addNewResolverAndRecords.apply(this, arguments);
            };

            addNewResolverAndRecords = function _ref6(_x4) {
              return _addNewResolverAndRecords.apply(this, arguments);
            };

            _addResolverAndRecords = function _ref5() {
              _addResolverAndRecords = (0, _asyncToGenerator2["default"])(
              /*#__PURE__*/
              _regenerator["default"].mark(function _callee2(name, resolverAddress) {
                var account,
                    hash,
                    _args2 = arguments;
                return _regenerator["default"].wrap(function _callee2$(_context2) {
                  while (1) {
                    switch (_context2.prev = _context2.next) {
                      case 0:
                        account = _args2.length > 2 && _args2[2] !== undefined ? _args2[2] : accounts[0];
                        console.log('Setting up ', name, 'with old resolver and records');
                        hash = namehash(name);
                        _context2.next = 5;
                        return ensContract.setResolver(hash, resolverAddress).send({
                          from: account
                        });

                      case 5:
                        _context2.next = 7;
                        return resolverContract.setAddr(hash, resolverAddress).send({
                          from: account
                        });

                      case 7:
                        _context2.next = 9;
                        return resolverContract.setContenthash(hash, contenthash).send({
                          gas: 5000000,
                          from: account
                        });

                      case 9:
                        console.log('finished setting up old resolver and records', name);

                      case 10:
                      case "end":
                        return _context2.stop();
                    }
                  }
                }, _callee2);
              }));
              return _addResolverAndRecords.apply(this, arguments);
            };

            addResolverAndRecords = function _ref4(_x2, _x3) {
              return _addResolverAndRecords.apply(this, arguments);
            };

            namehash = function _ref3(name) {
              var node = '0x0000000000000000000000000000000000000000000000000000000000000000';

              if (name !== '') {
                var _labels = name.split('.');

                for (var _i = _labels.length - 1; _i >= 0; _i--) {
                  node = sha3(node + sha3(_labels[_i]).slice(2), {
                    encoding: 'hex'
                  });
                }
              }

              return node.toString();
            };

            web3 = _ref.web3, accounts = _ref.accounts, _ref$dnssec = _ref.dnssec, dnssec = _ref$dnssec === void 0 ? false : _ref$dnssec;
            sha3 = web3.utils.sha3;
            nameLogger = new _namelogger.NameLogger({
              sha3: sha3,
              namehash: namehash
            });
            registryJSON = (0, _utils.loadContract)('registry', 'ENSRegistry');
            resolverJSON = (0, _utils.loadContract)('resolvers', 'PublicResolver');
            oldResolverJSON = (0, _utils.loadContract)('ens-022', 'PublicResolver');
            reverseRegistrarJSON = (0, _utils.loadContract)('registry', 'ReverseRegistrar');
            priceOracleJSON = (0, _utils.loadContract)('ethregistrar-202', 'SimplePriceOracle');
            linearPremiumPriceOracleJSON = (0, _utils.loadContract)('ethregistrar', 'LinearPremiumPriceOracle');
            dummyOracleJSON = (0, _utils.loadContract)('ethregistrar', 'DummyOracle');
            controllerJSON = (0, _utils.loadContract)('ethregistrar', 'ETHRegistrarController');
            bulkRenewalJSON = (0, _utils.loadContract)('ethregistrar', 'BulkRenewal');
            testRegistrarJSON = (0, _utils.loadContract)('registry', 'TestRegistrar');
            legacyAuctionRegistrarSimplifiedJSON = (0, _utils.loadContract)('ens-022', 'HashRegistrar');
            ENSWithFallbackJSON = (0, _utils.loadContract)('registry', 'ENSRegistryWithFallback');
            oldBaseRegistrarJSON = (0, _utils.loadContract)('ethregistrar-202', 'OldBaseRegistrarImplementation');
            newBaseRegistrarJSON = (0, _utils.loadContract)('ethregistrar', 'BaseRegistrarImplementation');
            registrarMigrationJSON = (0, _utils.loadContract)('ethregistrar-202', 'RegistrarMigration');
            EthRegistrarSubdomainRegistrarJSON = (0, _utils.loadContract)('subdomain-registrar', 'EthRegistrarSubdomainRegistrar');
            ENSMigrationSubdomainRegistrarJSON = (0, _utils.loadContract)('subdomain-registrar', 'ENSMigrationSubdomainRegistrar');
            console.log('Deploying from account ', accounts[0]);
            /* Deploy the main contracts  */

            _context5.prev = 28;
            _context5.next = 31;
            return (0, _utils.deploy)(web3, accounts[0], registryJSON);

          case 31:
            ens = _context5.sent;
            _context5.next = 34;
            return (0, _utils.deploy)(web3, accounts[0], resolverJSON, ens._address, ZERO_ADDRESS);

          case 34:
            resolver = _context5.sent;
            _context5.next = 37;
            return (0, _utils.deploy)(web3, accounts[0], oldResolverJSON, ens._address);

          case 37:
            oldResolver = _context5.sent;
            _context5.next = 40;
            return (0, _utils.deploy)(web3, accounts[0], reverseRegistrarJSON, ens._address, resolver._address);

          case 40:
            oldReverseRegistrar = _context5.sent;
            _context5.next = 43;
            return (0, _utils.deploy)(web3, accounts[0], testRegistrarJSON, ens._address, namehash('test'));

          case 43:
            testRegistrar = _context5.sent;
            eightweeks = 60 * 60 * 24 * 7 * 8;
            _context5.next = 47;
            return web3.eth.getBlock('latest');

          case 47:
            _context5.t0 = _context5.sent.timestamp;
            _context5.t1 = eightweeks;
            startTime = _context5.t0 - _context5.t1;
            _context5.next = 52;
            return (0, _utils.deploy)(web3, accounts[0], legacyAuctionRegistrarSimplifiedJSON, ens._address, namehash('eth'), startTime);

          case 52:
            legacyAuctionRegistrar = _context5.sent;
            _context5.next = 58;
            break;

          case 55:
            _context5.prev = 55;
            _context5.t2 = _context5["catch"](28);
            console.log('deployment failed', _context5.t2);

          case 58:
            ensContract = ens.methods;
            resolverContract = resolver.methods;
            oldResolverContract = oldResolver.methods;
            oldReverseRegistrarContract = oldReverseRegistrar.methods;
            testRegistrarContract = testRegistrar.methods;
            legacyAuctionRegistrarContract = legacyAuctionRegistrar.methods;
            tld = 'eth';
            tldHash = sha3(tld);
            /* Setup the root TLD */

            _context5.next = 68;
            return ensContract.setSubnodeOwner(ROOT_NODE, tldHash, accounts[0]).send({
              from: accounts[0]
            });

          case 68:
            _context5.next = 70;
            return ensContract.setSubnodeOwner(ROOT_NODE, sha3('test'), accounts[0]).send({
              from: accounts[0]
            });

          case 70:
            _context5.next = 72;
            return ensContract.setResolver(namehash(''), resolver._address).send({
              from: accounts[0]
            });

          case 72:
            _context5.next = 74;
            return ensContract.setResolver(namehash('eth'), resolver._address).send({
              from: accounts[0]
            });

          case 74:
            _context5.next = 76;
            return ensContract.setResolver(namehash('test'), resolver._address).send({
              from: accounts[0]
            });

          case 76:
            _context5.next = 78;
            return ensContract.setSubnodeOwner(ROOT_NODE, sha3('test'), testRegistrar._address).send({
              from: accounts[0]
            });

          case 78:
            _context5.next = 80;
            return ensContract.setSubnodeOwner(ROOT_NODE, sha3('eth'), legacyAuctionRegistrar._address).send({
              from: accounts[0]
            });

          case 80:
            legacynames = ['auctioned2', 'auctioned3'];

            if (!dnssec) {
              _context5.next = 85;
              break;
            }

            console.log('*** Skipping auction to make DNSSEC work');
            _context5.next = 103;
            break;

          case 85:
            _context5.prev = 85;
            i = 0;

          case 87:
            if (!(i < legacynames.length)) {
              _context5.next = 93;
              break;
            }

            _context5.next = 90;
            return (0, _utils.auctionLegacyName)(web3, accounts[0], legacyAuctionRegistrarContract, legacynames[i]);

          case 90:
            i++;
            _context5.next = 87;
            break;

          case 93:
            _context5.next = 98;
            break;

          case 95:
            _context5.prev = 95;
            _context5.t3 = _context5["catch"](85);
            console.log('auctioning Legacy name failed', {
              name: legacynames[i],
              e: _context5.t3
            });

          case 98:
            lockoutlength = 60 * 60 * 24 * 190;
            _context5.next = 101;
            return (0, _utils.advanceTime)(web3, lockoutlength);

          case 101:
            _context5.next = 103;
            return (0, _utils.mine)(web3);

          case 103:
            _context5.next = 105;
            return ensContract.setSubnodeOwner(ROOT_NODE, sha3('reverse'), accounts[0]).send({
              from: accounts[0]
            });

          case 105:
            nameLogger.record('reverse', {
              label: 'reverse'
            });
            _context5.next = 108;
            return ensContract.setSubnodeOwner(namehash('reverse'), sha3('addr'), accounts[0]).send({
              from: accounts[0]
            });

          case 108:
            console.log('setup root reverse with addr label');
            nameLogger.record('addr.reverse', {
              label: 'addr'
            });
            _context5.next = 112;
            return ensContract.setResolver(namehash('addr.reverse'), resolver._address).send({
              from: accounts[0]
            });

          case 112:
            console.log('setup root reverse with public resolver');
            /* Setup the reverse subdomain: addr.reverse */

            _context5.next = 115;
            return ensContract.setSubnodeOwner(namehash('reverse'), sha3('addr'), oldReverseRegistrar._address).send({
              from: accounts[0]
            });

          case 115:
            console.log('setup root reverse with the reverse registrar');
            /* Set the old hash registrar contract as the owner of .eth */

            _context5.next = 118;
            return ensContract.setSubnodeOwner(ROOT_NODE, tldHash, legacyAuctionRegistrar._address).send({
              from: accounts[0]
            });

          case 118:
            nameLogger.record('eth', {
              label: 'eth'
            });
            console.log('Successfully setup old hash registrar');
            _context5.next = 122;
            return web3.eth.getBlock('latest');

          case 122:
            now = _context5.sent.timestamp;
            _context5.next = 125;
            return (0, _utils.deploy)(web3, accounts[0], priceOracleJSON, 1);

          case 125:
            priceOracle = _context5.sent;
            _context5.next = 128;
            return (0, _utils.deploy)(web3, accounts[0], oldBaseRegistrarJSON, ens._address, legacyAuctionRegistrar._address, namehash('eth'), now + 365 * _utils.DAYS);

          case 128:
            oldBaseRegistrar = _context5.sent;
            console.log('Successfully setup base registrar');
            _context5.next = 132;
            return (0, _utils.deploy)(web3, accounts[0], controllerJSON, oldBaseRegistrar._address, priceOracle._address, 2, // 10 mins in seconds
            86400 // 24 hours in seconds
            );

          case 132:
            controller = _context5.sent;
            console.log('Successfully setup permanent registrar controller');
            oldBaseRegistrarContract = oldBaseRegistrar.methods;
            controllerContract = controller.methods;
            console.log('Price oracle deployed at: ', priceOracle._address);
            console.log('Base registrar deployed at: ', oldBaseRegistrar._address);
            console.log('Controller deployed at: ', controller._address);
            _context5.next = 141;
            return ensContract.setSubnodeOwner(ROOT_NODE, tldHash, accounts[0]).send({
              from: accounts[0]
            });

          case 141:
            _context5.prev = 141;
            _context5.next = 144;
            return resolverContract.setInterface(namehash('eth'), legacyRegistrarInterfaceId, legacyAuctionRegistrar._address).send({
              from: accounts[0]
            });

          case 144:
            _context5.next = 149;
            break;

          case 146:
            _context5.prev = 146;
            _context5.t4 = _context5["catch"](141);
            console.log(_context5.t4);

          case 149:
            console.log("Set .eth legacy registrar interface Id to ".concat(legacyAuctionRegistrar._address));
            _context5.next = 152;
            return resolverContract.setInterface(namehash('eth'), permanentRegistrarInterfaceId, controller._address).send({
              from: accounts[0]
            });

          case 152:
            console.log("Set .eth permanent registrar interface Id to ".concat(controller._address));
            /* Set the permanent registrar contract as the owner of .eth */

            _context5.next = 155;
            return ensContract.setSubnodeOwner(ROOT_NODE, tldHash, oldBaseRegistrar._address).send({
              from: accounts[0]
            });

          case 155:
            console.log('Add controller to base registrar');
            _context5.next = 158;
            return oldBaseRegistrarContract.addController(controller._address).send({
              from: accounts[0]
            });

          case 158:
            newnames = ['testing', 'newname', 'resolver', 'oldresolver', 'awesome', 'superawesome', 'notsoawesome', 'abittooawesome', 'abittooawesome2', 'abittooawesome3', 'subdomaindummy', 'contractdomain'];
            console.log('Register name');
            _context5.prev = 160;
            i = 0;

          case 162:
            if (!(i < newnames.length)) {
              _context5.next = 169;
              break;
            }

            _context5.next = 165;
            return (0, _utils.registerName)(web3, accounts[0], controllerContract, newnames[i]);

          case 165:
            nameLogger.record("".concat(newnames[i], ".eth"), {
              label: newnames[i]
            });

          case 166:
            i++;
            _context5.next = 162;
            break;

          case 169:
            _context5.next = 174;
            break;

          case 171:
            _context5.prev = 171;
            _context5.t5 = _context5["catch"](160);
            console.log('Failed to register a name', _context5.t5);

          case 174:
            _context5.next = 176;
            return (0, _utils.registerName)(web3, accounts[1], controllerContract, 'otherowner');

          case 176:
            nameLogger.record("otherowner.eth", {
              label: 'otherowner'
            });
            newnames.push('otherowner');
            /* Setup domain with a resolver and addr/content */

            aBitTooAwesome = 'abittooawesome.eth';
            aBitTooAwesome2 = 'abittooawesome2.eth';
            aBitTooAwesome3 = 'abittooawesome3.eth';
            otherOwner = 'otherowner.eth';
            addResolverAndRecords(aBitTooAwesome2, resolver._address);
            addResolverAndRecords(aBitTooAwesome3, resolver._address);
            addResolverAndRecords(otherOwner, resolver._address, accounts[1]);
            contractdomain = namehash('contractdomain.eth');
            _context5.next = 188;
            return ensContract.setResolver(contractdomain, resolver._address).send({
              from: accounts[0]
            });

          case 188:
            _context5.next = 190;
            return resolverContract.setAddr(contractdomain, accounts[0]).send({
              from: accounts[0]
            });

          case 190:
            _context5.next = 192;
            return ensContract.setOwner(contractdomain, testRegistrar._address).send({
              from: accounts[0]
            });

          case 192:
            _context5.next = 194;
            return oldReverseRegistrarContract.setName('abittooawesome.eth').send({
              from: accounts[0],
              gas: 1000000
            });

          case 194:
            /* Point the resolver.eth's resolver to the public resolver */
            console.log('Setting up resolvers');
            _context5.next = 197;
            return ensContract.setResolver(namehash('resolver.eth'), resolver._address).send({
              from: accounts[0]
            });

          case 197:
            _context5.next = 199;
            return ensContract.setResolver(namehash('oldresolver.eth'), oldResolver._address).send({
              from: accounts[0]
            });

          case 199:
            _context5.next = 201;
            return resolverContract.setAddr(namehash('resolver.eth'), resolver._address).send({
              from: accounts[0]
            });

          case 201:
            _context5.next = 203;
            return resolverContract.setAddr(namehash('oldresolver.eth'), oldResolver._address).send({
              from: accounts[0]
            });

          case 203:
            /* Resolve the resolver.eth content to a 32 byte content hash */
            console.log('Setting up contenthash');
            _context5.next = 206;
            return resolverContract.setContenthash(namehash('resolver.eth'), contenthash).send({
              from: accounts[0],
              gas: 5000000
            });

          case 206:
            _context5.next = 208;
            return oldResolverContract.setContent(namehash('oldresolver.eth'), content).send({
              from: accounts[0]
            });

          case 208:
            _context5.next = 210;
            return oldReverseRegistrarContract.setName('eth').send({
              from: accounts[2],
              gas: 1000000
            });

          case 210:
            _context5.next = 212;
            return (0, _utils.mine)(web3);

          case 212:
            _context5.next = 214;
            return web3.eth.getBlock('latest');

          case 214:
            current = _context5.sent;
            console.log("The current time is ".concat(new Date(current.timestamp * 1000)));
            oldEns = ens;
            label = 'notmigrated';
            _context5.next = 220;
            return (0, _utils.registerName)(web3, accounts[0], controllerContract, label);

          case 220:
            nameLogger.record("".concat(label, ".eth"), {
              label: label
            });
            _context5.next = 223;
            return ensContract.setSubnodeOwner(namehash('testing.eth'), sha3('sub1'), accounts[0]).send({
              from: accounts[0]
            });

          case 223:
            nameLogger.record("sub1.testing.eth", {
              label: 'sub1'
            });
            _context5.next = 226;
            return ensContract.setSubnodeOwner(namehash('testing.eth'), sha3('sub2'), accounts[0]).send({
              from: accounts[0]
            });

          case 226:
            _context5.next = 228;
            return ensContract.setResolver(namehash('sub1.testing.eth'), resolver._address).send({
              from: accounts[0]
            });

          case 228:
            _context5.next = 230;
            return ensContract.setResolver(namehash('sub2.testing.eth'), resolver._address).send({
              from: accounts[0]
            });

          case 230:
            _context5.next = 232;
            return resolverContract.setAddr(namehash('sub2.testing.eth'), accounts[0]).send({
              from: accounts[0]
            });

          case 232:
            _context5.next = 234;
            return resolverContract['setAddr(bytes32,uint256,bytes)'](namehash('sub2.testing.eth'), 0, //BTC
            '0x76a91462e907b15cbf27d5425399ebf6f0fb50ebb88f1888ac' //1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa
            ).send({
              from: accounts[0]
            });

          case 234:
            _context5.next = 236;
            return resolverContract.setContenthash(namehash('sub2.testing.eth'), contenthash).send({
              gas: 5000000,
              from: accounts[0]
            });

          case 236:
            nameLogger.record("sub2.testing.eth", {
              label: 'sub2'
            });
            _context5.next = 239;
            return ensContract.setSubnodeOwner(namehash('sub2.testing.eth'), sha3('a1'), accounts[0]).send({
              from: accounts[0]
            });

          case 239:
            _context5.next = 241;
            return ensContract.setResolver(namehash('a1.sub2.testing.eth'), resolver._address).send({
              from: accounts[0]
            });

          case 241:
            _context5.next = 243;
            return resolverContract.setAddr(namehash('a1.sub2.testing.eth'), accounts[0]).send({
              from: accounts[0]
            });

          case 243:
            nameLogger.record("a1.sub2.testing.eth", {
              label: 'a1'
            });
            _context5.next = 246;
            return ensContract.setSubnodeOwner(namehash('otherowner.eth'), sha3('sub1'), accounts[0]).send({
              from: accounts[1]
            });

          case 246:
            _context5.next = 248;
            return ensContract.setResolver(namehash('sub1.otherowner.eth'), resolver._address).send({
              from: accounts[0]
            });

          case 248:
            _context5.next = 250;
            return resolverContract.setAddr(namehash('sub1.otherowner.eth'), accounts[0]).send({
              from: accounts[0]
            });

          case 250:
            nameLogger.record("sub1.otherowner.eth", {
              label: 'sub1'
            });
            _context5.next = 253;
            return ensContract.setSubnodeOwner(namehash('otherowner.eth'), sha3('sub2'), accounts[1]).send({
              from: accounts[1]
            });

          case 253:
            _context5.next = 255;
            return ensContract.setResolver(namehash('sub2.otherowner.eth'), resolver._address).send({
              from: accounts[1]
            });

          case 255:
            _context5.next = 257;
            return resolverContract.setAddr(namehash('sub2.otherowner.eth'), accounts[1]).send({
              from: accounts[1]
            });

          case 257:
            nameLogger.record("sub2.otherowner.eth", {
              label: 'sub2'
            });
            _context5.next = 260;
            return ensContract.setSubnodeOwner(namehash('testing.eth'), sha3('sub4'), accounts[1]).send({
              from: accounts[0]
            });

          case 260:
            _context5.next = 262;
            return ensContract.setResolver(namehash('sub4.testing.eth'), resolver._address).send({
              from: accounts[1]
            });

          case 262:
            _context5.next = 264;
            return resolverContract.setAddr(namehash('sub4.testing.eth'), accounts[0]).send({
              from: accounts[1]
            });

          case 264:
            _context5.next = 266;
            return resolverContract['setAddr(bytes32,uint256,bytes)'](namehash('sub4.testing.eth'), 0, //BTC
            '0x76a91462e907b15cbf27d5425399ebf6f0fb50ebb88f1888ac' //1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa
            ).send({
              from: accounts[1]
            });

          case 266:
            nameLogger.record("sub4.testing.eth", {
              label: 'sub4'
            });
            _context5.next = 269;
            return (0, _utils.deploy)(web3, accounts[0], EthRegistrarSubdomainRegistrarJSON, ens._address);

          case 269:
            oldSubdomainRegistrar = _context5.sent;
            oldSubdomainRegistrarContract = oldSubdomainRegistrar.methods; // Create the new subdomain registrar

            _context5.next = 273;
            return (0, _utils.deploy)(web3, accounts[0], ENSMigrationSubdomainRegistrarJSON, ens._address);

          case 273:
            subdomainRegistrar = _context5.sent;
            _context5.next = 276;
            return (0, _utils.deploy)(web3, accounts[0], ENSWithFallbackJSON, oldEns._address);

          case 276:
            newEns = _context5.sent;
            newEnsContract = newEns.methods;
            _context5.next = 280;
            return (0, _utils.deploy)(web3, accounts[0], newBaseRegistrarJSON, newEns._address, namehash('eth'));

          case 280:
            newBaseRegistrar = _context5.sent;
            newBaseRegistrarContract = newBaseRegistrar.methods;
            _context5.next = 284;
            return newBaseRegistrarContract.addController(accounts[0]).send({
              from: accounts[0]
            });

          case 284:
            // Create the new controller
            // Dummy oracle with 1 ETH == 200 USD
            dummyOracleRate = toBN(20000000000); // 200 * 1e8

            _context5.next = 287;
            return (0, _utils.deploy)(web3, accounts[0], dummyOracleJSON, dummyOracleRate);

          case 287:
            dummyOracle = _context5.sent;
            dummyOracleContract = dummyOracle.methods;
            _context5.next = 291;
            return dummyOracleContract.latestAnswer().call();

          case 291:
            latestAnswer = _context5.sent;
            console.log('Dummy USD Rate', {
              latestAnswer: latestAnswer
            }); // Premium starting price: 10 ETH = 2000 USD

            premium = toBN('2000000000000000000000'); // 2000 * 1e18

            decreaseDuration = toBN(28 * _utils.DAYS);
            decreaseRate = premium.div(decreaseDuration);
            _context5.next = 298;
            return (0, _utils.deploy)(web3, accounts[0], linearPremiumPriceOracleJSON, dummyOracle._address, // Oracle prices from https://etherscan.io/address/0xb9d374d0fe3d8341155663fae31b7beae0ae233a#events
            // 0,0, 127, 32, 1
            [0, 0, toBN(20294266869609), toBN(5073566717402), toBN(158548959919)], premium, decreaseRate);

          case 298:
            linearPriceOracle = _context5.sent;
            linearPriceOracleContract = linearPriceOracle.methods;
            _context5.next = 302;
            return (0, _utils.deploy)(web3, accounts[0], controllerJSON, newBaseRegistrar._address, linearPriceOracle._address, 2, // 10 mins in seconds
            86400 // 24 hours in seconds
            );

          case 302:
            newController = _context5.sent;
            newControllerContract = newController.methods; // Create the new resolver

            _context5.next = 306;
            return (0, _utils.deploy)(web3, accounts[0], resolverJSON, newEns._address, ZERO_ADDRESS);

          case 306:
            newResolver = _context5.sent;
            newResolverContract = newResolver.methods; // Set resolver to the new ENS

            _context5.next = 310;
            return (0, _utils.deploy)(web3, accounts[0], bulkRenewalJSON, newEns._address);

          case 310:
            bulkRenewal = _context5.sent;

            if (!dnssec) {
              _context5.next = 314;
              break;
            }

            _context5.next = 314;
            return (0, _deployDNSSEC["default"])(web3, accounts, newEns, newResolver);

          case 314:
            _context5.next = 316;
            return newEnsContract.setSubnodeOwner(ROOT_NODE, sha3('eth'), accounts[0]).send({
              from: accounts[0]
            });

          case 316:
            _context5.next = 318;
            return newEnsContract.setResolver(namehash('eth'), newResolver._address).send({
              from: accounts[0],
              gas: 6000000
            });

          case 318:
            _context5.next = 320;
            return newResolverContract.setApprovalForAll(newController._address, true).send({
              from: accounts[0]
            });

          case 320:
            _context5.next = 322;
            return newResolverContract.setInterface(namehash('eth'), permanentRegistrarInterfaceId, newController._address).send({
              from: accounts[0]
            });

          case 322:
            _context5.next = 324;
            return newResolverContract.setInterface(namehash('eth'), permanentRegistrarWithConfigInterfaceId, newController._address).send({
              from: accounts[0]
            });

          case 324:
            if (dnssec) {
              _context5.next = 327;
              break;
            }

            _context5.next = 327;
            return newResolverContract.setInterface(namehash('eth'), legacyRegistrarInterfaceId, legacyAuctionRegistrar._address).send({
              from: accounts[0]
            });

          case 327:
            _context5.next = 329;
            return newResolverContract.setInterface(namehash('eth'), bulkRenewalInterfaceId, bulkRenewal._address).send({
              from: accounts[0]
            });

          case 329:
            _context5.next = 331;
            return newResolverContract.setInterface(namehash('eth'), linearPriceOracleInterfaceId, linearPriceOracle._address).send({
              from: accounts[0]
            });

          case 331:
            _context5.next = 333;
            return newEnsContract.setSubnodeOwner(ROOT_NODE, sha3('eth'), newBaseRegistrar._address).send({
              from: accounts[0]
            });

          case 333:
            nameLogger.record('eth', {
              label: 'eth',
              migrated: true
            });
            _context5.next = 336;
            return (0, _utils.deploy)(web3, accounts[0], testRegistrarJSON, newEns._address, namehash('test'));

          case 336:
            newTestRegistrar = _context5.sent;
            newTestRegistrarContract = newTestRegistrar.methods;
            _context5.next = 340;
            return newEnsContract.setSubnodeOwner(ROOT_NODE, sha3('test'), newTestRegistrar._address).send({
              from: accounts[0]
            });

          case 340:
            nameLogger.record('test', {
              label: 'test',
              migrated: true
            });
            _context5.next = 343;
            return (0, _utils.deploy)(web3, accounts[0], reverseRegistrarJSON, newEns._address, newResolver._address);

          case 343:
            newReverseRegistrar = _context5.sent;
            _context5.next = 346;
            return (0, _utils.deploy)(web3, accounts[0], registrarMigrationJSON, oldBaseRegistrar._address, newBaseRegistrar._address, oldSubdomainRegistrar._address, subdomainRegistrar._address);

          case 346:
            registrarMigration = _context5.sent;
            registrarMigrationContract = registrarMigration.methods;
            _context5.next = 350;
            return newBaseRegistrarContract.addController(registrarMigration._address).send({
              from: accounts[0]
            });

          case 350:
            _context5.next = 352;
            return ensContract.setSubnodeOwner(ROOT_NODE, sha3('eth'), registrarMigration._address).send({
              from: accounts[0]
            });

          case 352:
            console.log('Migrating permanent registrar names');
            _context5.prev = 353;
            i = 0;

          case 355:
            if (!(i < newnames.length)) {
              _context5.next = 377;
              break;
            }

            name = newnames[i];
            domain = "".concat(name, ".eth");
            labelhash = sha3(name);
            nameLogger.record(domain, {
              label: name,
              migrated: true
            });
            _context5.next = 362;
            return ensContract.owner(namehash(domain));

          case 362:
            owner = _context5.sent;

            if (!(owner === accounts[0])) {
              _context5.next = 370;
              break;
            }

            _context5.next = 366;
            return ensContract.setTTL(namehash(domain), 123).send({
              from: accounts[0]
            });

          case 366:
            _context5.next = 368;
            return ensContract.setResolver(namehash(domain), newResolver._address).send({
              from: accounts[0]
            });

          case 368:
            _context5.next = 371;
            break;

          case 370:
            console.log("".concat(domain, " is not owned by ").concat(accounts[0], " hence not setting ttl nor resolver"));

          case 371:
            _context5.next = 373;
            return registrarMigrationContract.migrate(labelhash).send({
              from: accounts[0],
              gas: 6000000
            });

          case 373:
            tx = _context5.sent;

          case 374:
            i++;
            _context5.next = 355;
            break;

          case 377:
            _context5.next = 382;
            break;

          case 379:
            _context5.prev = 379;
            _context5.t6 = _context5["catch"](353);
            console.log('Failed to migrate a name', _context5.t6);

          case 382:
            if (dnssec) {
              _context5.next = 402;
              break;
            }

            console.log('Migrate legacy names');
            _context5.prev = 384;

            migrate =
            /*#__PURE__*/
            function () {
              var _ref2 = (0, _asyncToGenerator2["default"])(
              /*#__PURE__*/
              _regenerator["default"].mark(function _callee(label) {
                var name, domain, labelhash, tx;
                return _regenerator["default"].wrap(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        name = label;
                        domain = "".concat(name, ".eth");
                        labelhash = sha3(name);
                        console.log("Migrate legacy ".concat(domain));
                        _context.next = 6;
                        return ensContract.setResolver(namehash(domain), resolver._address).send({
                          from: accounts[0]
                        });

                      case 6:
                        _context.next = 8;
                        return ensContract.setTTL(namehash(domain), 123).send({
                          from: accounts[0]
                        });

                      case 8:
                        _context.next = 10;
                        return registrarMigrationContract.migrateLegacy(labelhash).send({
                          from: accounts[0],
                          gas: 6000000
                        });

                      case 10:
                        tx = _context.sent;

                      case 11:
                      case "end":
                        return _context.stop();
                    }
                  }
                }, _callee);
              }));

              return function migrate(_x5) {
                return _ref2.apply(this, arguments);
              };
            }();

            i = 0;

          case 387:
            if (!(i < legacynames.length)) {
              _context5.next = 394;
              break;
            }

            _context5.next = 390;
            return migrate(legacynames[i]);

          case 390:
            nameLogger.record("".concat(legacynames[i], ".eth"), {
              label: legacynames[i],
              migrated: true
            });

          case 391:
            i++;
            _context5.next = 387;
            break;

          case 394:
            _context5.next = 399;
            break;

          case 396:
            _context5.prev = 396;
            _context5.t7 = _context5["catch"](384);
            console.log('Failed to migrate a name', _context5.t7);

          case 399:
            console.log("Releasing the deed of auctioned2");
            _context5.next = 402;
            return legacyAuctionRegistrarContract.releaseDeed(sha3('auctioned2')).send({
              from: accounts[0]
            });

          case 402:
            _context5.next = 404;
            return newEnsContract.setResolver(namehash('resolver.eth'), newResolver._address).send({
              from: accounts[0]
            });

          case 404:
            console.log('Set resolver.eth address to new resovler address', newResolver._address);
            _context5.next = 407;
            return newResolverContract.setAddr(namehash('resolver.eth'), newResolver._address).send({
              from: accounts[0]
            });

          case 407:
            _context5.next = 409;
            return newEnsContract.setSubnodeOwner(namehash('testing.eth'), sha3('sub3'), accounts[0]).send({
              from: accounts[0]
            });

          case 409:
            nameLogger.record('testing.eth', {
              label: 'sub3',
              migrated: true
            });
            _context5.next = 412;
            return setNewResolver('notsoawesome.eth');

          case 412:
            _context5.next = 414;
            return addNewResolverAndRecords('abittooawesome.eth');

          case 414:
            /* Setup some domains for subdomain testing */
            console.log('Setting up subdomaindummy.eth');
            _context5.next = 417;
            return newEnsContract.setSubnodeOwner(namehash('subdomaindummy.eth'), sha3('original'), accounts[0]).send({
              from: accounts[0]
            });

          case 417:
            // Change the controller from migration registrarMigration to controller
            nameLogger.record('original.subdomaindummy.eth', {
              label: 'original',
              migrated: true
            });
            console.log("Add Controller ".concat(newController._address, "  to new base registrar"));
            _context5.next = 421;
            return newBaseRegistrarContract.addController(newController._address).send({
              from: accounts[0]
            });

          case 421:
            _context5.next = 423;
            return (0, _utils.registerName)(web3, accounts[0], newControllerContract, 'aftermigration');

          case 423:
            _context5.next = 425;
            return (0, _utils.registerName)(web3, accounts[0], newControllerContract, 'postmigration');

          case 425:
            nameLogger.record('postmigration.eth', {
              label: 'postmigration',
              migrated: true
            });
            _context5.next = 428;
            return newEnsContract.setSubnodeOwner(ROOT_NODE, sha3('reverse'), accounts[1]).send({
              from: accounts[0]
            });

          case 428:
            nameLogger.record('reverse', {
              label: 'reverse',
              migrated: true
            });
            _context5.next = 431;
            return newEnsContract.setSubnodeOwner(namehash('reverse'), sha3('addr'), newReverseRegistrar._address).send({
              from: accounts[1]
            });

          case 431:
            nameLogger.record('addr.reverse', {
              label: 'addr',
              migrated: true
            });
            _context5.next = 434;
            return newTestRegistrarContract.register(sha3('example'), accounts[0]).send({
              from: accounts[0]
            });

          case 434:
            nameLogger.record('example.test', {
              label: 'example',
              migrated: true
            });
            baseDays = 60;
            _context5.t8 = Date;
            _context5.next = 439;
            return web3.eth.getBlock('latest');

          case 439:
            _context5.t9 = _context5.sent.timestamp;
            _context5.t10 = _context5.t9 * 1000;
            beforeTime = new _context5.t8(_context5.t10);
            _context5.next = 444;
            return (0, _utils.registerName)(web3, accounts[0], newControllerContract, 'justreleased', baseDays * _utils.DAYS);

          case 444:
            _context5.next = 446;
            return (0, _utils.registerName)(web3, accounts[0], newControllerContract, 'released', (baseDays - 15) * _utils.DAYS);

          case 446:
            nameLogger.record('released', {
              label: 'released',
              migrated: true
            });
            _context5.next = 449;
            return (0, _utils.registerName)(web3, accounts[0], newControllerContract, 'rele', (baseDays - 15) * _utils.DAYS);

          case 449:
            nameLogger.record('rele', {
              label: 'rele',
              migrated: true
            });
            _context5.next = 452;
            return (0, _utils.registerName)(web3, accounts[0], newControllerContract, 'rel', (baseDays - 15) * _utils.DAYS);

          case 452:
            nameLogger.record('rel', {
              label: 'rel',
              migrated: true
            });
            _context5.next = 455;
            return (0, _utils.registerName)(web3, accounts[0], newControllerContract, 'onedaypremium', (baseDays - 27) * _utils.DAYS);

          case 455:
            nameLogger.record('onedaypremium', {
              label: 'onedaypremium',
              migrated: true
            });
            nameLogger.record('justreleased', {
              label: 'justreleased',
              migrated: true
            });

            if (dnssec) {
              _context5.next = 462;
              break;
            }

            _context5.next = 460;
            return (0, _utils.advanceTime)(web3, (baseDays + 90) * _utils.DAYS + 1);

          case 460:
            _context5.next = 462;
            return (0, _utils.mine)(web3);

          case 462:
            _context5.t11 = Date;
            _context5.next = 465;
            return web3.eth.getBlock('latest');

          case 465:
            _context5.t12 = _context5.sent.timestamp;
            _context5.t13 = _context5.t12 * 1000;
            afterTime = new _context5.t11(_context5.t13);
            console.log({
              beforeTime: beforeTime,
              afterTime: afterTime
            });
            _context5.next = 471;
            return linearPriceOracleContract.price('somenonexsitingname122', 0, 12 * 30 * _utils.DAYS).call();

          case 471:
            sixcharprice = _context5.sent;
            _context5.next = 474;
            return linearPriceOracleContract.price('1234', 0, 12 * 30 * _utils.DAYS).call();

          case 474:
            fourcharprice = _context5.sent;
            _context5.next = 477;
            return linearPriceOracleContract.price('123', 0, 12 * 30 * _utils.DAYS).call();

          case 477:
            threecharprice = _context5.sent;
            console.log({
              sixcharprice: sixcharprice,
              fourcharprice: fourcharprice,
              threecharprice: threecharprice
            });
            _context5.next = 481;
            return newEnsContract.setResolver(namehash('abittooawesome2.eth'), newResolver._address).send({
              from: accounts[0]
            });

          case 481:
            _context5.next = 483;
            return newResolverContract.setContenthash(namehash('abittooawesome2.eth'), deprecated_contenthash).send({
              from: accounts[0]
            });

          case 483:
            // Disabled for now as configureDomain is throwing errorr
            // await subdomainRegistrarContract.migrateSubdomain(namehash.hash("ismoney.eth"), sha3("eth")).send({from: accounts[0]})
            response = (_response = {
              emptyAddress: '0x0000000000000000000000000000000000000000',
              ownerAddress: accounts[0],
              bulkRenewalAddress: bulkRenewal._address,
              legacyAuctionRegistrarAddress: legacyAuctionRegistrar._address,
              oldEnsAddress: ens._address,
              oldContentResolverAddresses: [oldResolver._address],
              oldResolverAddresses: [resolver._address, oldResolver._address],
              oldControllerAddress: controller._address,
              oldBaseRegistrarAddress: oldBaseRegistrar._address,
              reverseRegistrarAddress: oldReverseRegistrar._address,
              ensAddress: newEns._address,
              registrarMigration: registrarMigration && registrarMigration._address,
              resolverAddress: newResolver._address
            }, (0, _defineProperty2["default"])(_response, "reverseRegistrarAddress", newReverseRegistrar && newReverseRegistrar._address), (0, _defineProperty2["default"])(_response, "reverseRegistrarOwnerAddress", accounts[0]), (0, _defineProperty2["default"])(_response, "controllerAddress", newController._address), (0, _defineProperty2["default"])(_response, "baseRegistrarAddress", newBaseRegistrar._address), (0, _defineProperty2["default"])(_response, "linearPriceOracle", linearPriceOracle._address), (0, _defineProperty2["default"])(_response, "dummyOracle", dummyOracle._address), _response);
            config = {
              columns: {
                0: {
                  alignment: 'left',
                  width: 30
                },
                1: {
                  alignment: 'center',
                  width: 50
                }
              }
            };
            contractNames = Object.keys(response);
            contractAddressesTable = contractNames.map(function (key) {
              return [key, response[key]];
            });
            output = (0, _table.table)(contractAddressesTable, config);
            console.log('Deployed contracts');
            console.log(output);
            console.log('Names');
            labels = nameLogger.generateTable();
            response.labels = {};
            labels.map(function (l, i) {
              return i !== 0 ? response.labels[l[3].slice(2)] = l[2] : null;
            });
            console.log(nameLogger.print());
            return _context5.abrupt("return", response);

          case 496:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5, null, [[28, 55], [85, 95], [141, 146], [160, 171], [353, 379], [384, 396]]);
  }));
  return _deployENS.apply(this, arguments);
}

var _default = deployENS;
exports["default"] = _default;
